### PR TITLE
OIDC Auth

### LINKED ISSUE #46: Implement oauth/oidc auth in LISA
Foreman and hammer supports oauth/oidc authentication using Keycloak (or any other Idp for that matter). See https://www.theforeman.org/manuals/nightly/index.html#5.7.7SingleSign-onforForemanusingOpenIDConnectprotocol

We should support oauth/oidc auth here as well.

### PR DESCRIPTION
**Pull request by kamils-iRonin**
Closes #46

### CODE REVIEW

**Comment by laugmanuel** [comment]
The login itself seems to work!

A few notes from my side:
- Settings validation does not seem to work quite right, as I haven't configured anything but get:
  ```sh
  ❯ npx ajv -d src/settings/development.json -s src/settings/settings.schema.json
  src/settings/development.json valid
  ```
  But visiting the UI results in `TypeError: Cannot read properties of undefined (reading 'clientId')`
- The login works even without a valid Foreman instance (404 being returned) -> not sure how to handle that
- There two multiple logout options:
  image
- When changing the Foreman config to break during JWT validation, I see my username in LISA (top right), but stay on the login form. There is no error visible in the UI (but console tells me 401).

File: src/containers/LoginContainer/index.js [hunk 1]
```diff
@@ -1,106 +1,32 @@
...
-          </Card>
-        </Col>
-      </Row>
-    </CardGrid>
-  )
-}
+const LoginContainer = () => (
+  <Card isCompact isFlat className={css(sizing.w_50, spacing.mxAuto, spacing.pXl)}>
+    <CardTitle>
+      <Brand src={logo} alt="LISA" className={css(sizing.w_100, alignment.textAlignCenter)} />
```
Comment (laugmanuel) [inline src/containers/LoginContainer/index.js:15]: What about making this logo a little bit smaller (`w_50`?) to not hide the button in the login form for some specific resolutions?

File: src/containers/LoginContainer/index.js [hunk 2]
```diff
@@ -1,106 +1,32 @@
...
+    </CardTitle>
+    <CardBody>
+      <Flex direction={{ default: 'column' }}>
+        <FlexItem>
+          <ForemanAuth />
```
Comment (laugmanuel) [inline src/containers/LoginContainer/index.js:20]: what about making this also optional so a admin could disable regular foreman auth with a setting?

File: src/hooks/useUser.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+export function useOidcUser() {
+  const { userData, signOut } = useOidcAuth();
+
+  if(!isEnabled) {
+    return null
+  }
+
+  return userData ? {
+    username: get(userData, 'profile.preferred_username', null),
```
Comment (laugmanuel) [inline src/hooks/useUser.js:23]: the `profile.preferred_username` field should also be configurable with a setting. Sadly not all IdP's follow the same scheme or are configured the same way.

**Comment by laugmanuel** [comment]
I finally was able to test this implementation and left a few comments inline.

Additionally, I have two points:
- WDYT about sending some sort of dummy API request (status or current user endpoint?) to see if the login was indeed successful (and API is reachable)? I think we should drop the user back to the login form and show an error in that case instead of showing a successful login with `Sorry but something went wrong. Try again later!`.
- Refreshing the page while being logged in causes the spinner to appear and never finish - the session is basically broken then

Otherwise, login and logout seems to work fine.

**Review by laugmanuel — COMMENTED** [review]
I finally was able to test this implementation and left a few comments inline.

Additionally, I have two points:

WDYT about sending some sort of dummy API request (status or current user endpoint?) to see if the login was indeed successful (and API is reachable)? I think we should drop the user back to the login form and show an error in that case instead of showing a successful login with Sorry but something went wrong. Try again later!.
Refreshing the page while being logged in causes the spinner to appear and never finish - the session is basically broken then
Otherwise, login and logout seems to work fine.

**Comment by kamils-iRonin** [comment]
@laugmanuel What do you think about moving it to a plugin? It would require more effort, but we would be less dependent on `oidc-react`

**Comment by laugmanuel** [comment]
> @laugmanuel What do you think about moving it to a plugin? It would require more effort, but we would be less dependent on oidc-react

To be honest, I don't really see the benefit. I think OIDC will become more popular in the future and Foreman contains it in core, so I guess we should keep it in LISA itself, too. @timogoebel WDYT?

**Comment by timogoebel** [comment]
> @timogoebel WDYT?

I agree. As it's a Foreman core feature, I don't see much benefit in having the functionality in a plugin for LISA.

File: src/lib/foremanApi.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
+export const status = async (token) => {
+  const url = `${process.env.REACT_APP_FOREMAN_URL}/api/v2/status/`
+  const headers = {
+    'Authorization': `Bearer ${token}`
```
[the remark this answers is not in the document]
Comment (laugmanuel) [reply src/lib/foremanApi.js:4]: Using oauth `access_token` in `Authorization` header should work as expected, yes.

**Review by laugmanuel — APPROVED** [review]
All manual tests were successful!

**Comment by laugmanuel** [comment]
Thanks @kamils-iRonin !

**Pull request merged**